File: frontend/src/lib/api/studio/flow-diagrams.ts
/**
 * Flow Diagrams API
 * Educational Note: Handles AI-generated Mermaid diagrams for visualizing
 * processes, workflows, relationships, and more.
 */

import axios from 'axios';
import { API_BASE_URL } from '../client';
import type { JobStatus } from './index';
import { createLogger } from '@/lib/logger';

const log = createLogger('studio-flow-diagrams-api');

/**
 * Mermaid diagram types supported
 */
export type DiagramType =
  | 'flowchart'
  | 'sequence'
  | 'state'
  | 'er'
  | 'class'
  | 'pie'
  | 'gantt'
  | 'journey'
  | 'mindmap';

/**
 * Flow diagram job record from the API
 */
export interface FlowDiagramJob {
  id: string;
  source_id: string;
  source_name: string;
  direction: string;
  status: JobStatus;
  progress: string;
  error: string | null;
  mermaid_syntax: string | null;
  diagram_type: DiagramType | null;
  title: string | null;
  description: string | null;
  generation_time_seconds: number | null;
  // Edit lineage
  parent_job_id: string | null;
  edit_instructions: string | null;
  created_at: string;
  started_at: string | null;
  completed_at: string | null;
}

/**
 * Response from starting flow diagram generation
 */
export interface StartFlowDiagramResponse {
  success: boolean;
  job_id?: string;
  message?: string;
  source_name?: string;
  error?: string;
}

/**
 * Response from getting flow diagram job status
 */
export interface FlowDiagramJobStatusResponse {
  success: boolean;
  job?: FlowDiagramJob;
  error?: string;
}

/**
 * Response from listing flow diagram jobs
 */
export interface ListFlowDiagramJobsResponse {
  success: boolean;
  jobs: FlowDiagramJob[];
  count: number;
  error?: string;
}

/**
 * Flow Diagrams API
 */
export const flowDiagramsAPI = {
  /**
   * Start flow diagram generation or edit
   */
  async startGeneration(
    projectId: string,
    sourceId?: string,
    direction?: string,
    parentJobId?: string,
    editInstructions?: string
  ): Promise<StartFlowDiagramResponse> {
    try {
      const body: Record<string, unknown> = {
        direction: direction || 'Create a diagram showing the key processes and relationships.',
      };
      if (sourceId) body.source_id = sourceId;
      if (parentJobId) body.parent_job_id = parentJobId;
      if (editInstructions) body.edit_instructions = editInstructions;

      const response = await axios.post(
        `${API_BASE_URL}/projects/${projectId}/studio/flow-diagram`,
        body
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to start flow diagram generation');
      throw error;
    }
  },

  /**
   * Get the status of a flow diagram job
   */
  async getJobStatus(projectId: string, jobId: string): Promise<FlowDiagramJobStatusResponse> {
    try {
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/flow-diagram-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to get flow diagram job status');
      throw error;
    }
  },

  /**
   * List all flow diagram jobs for a project
   */
  async listJobs(projectId: string, sourceId?: string): Promise<ListFlowDiagramJobsResponse> {
    try {
      const params = sourceId ? { source_id: sourceId } : {};
      const response = await axios.get(
        `${API_BASE_URL}/projects/${projectId}/studio/flow-diagram-jobs`,
        { params }
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to list flow diagram jobs');
      throw error;
    }
  },

  /**
   * Poll flow diagram job status until complete or error
   *
   * Educational Note: Includes initial retry tolerance to handle timing
   * between job creation and first poll availability.
   */
  async pollJobStatus(
    projectId: string,
    jobId: string,
    onProgress?: (job: FlowDiagramJob) => void,
    intervalMs: number = 2000,
    maxAttempts: number = 60
  ): Promise<FlowDiagramJob> {
    let attempts = 0;
    let currentInterval = intervalMs;
    let initialRetries = 0;
    const maxInitialRetries = 3;

    while (attempts < maxAttempts) {
      const response = await this.getJobStatus(projectId, jobId);

      // Handle initial 404s with retry tolerance
      if (!response.success || !response.job) {
        if (initialRetries < maxInitialRetries) {
          log.debug(`flow diagram job not found yet, retrying... (${initialRetries + 1}/${maxInitialRetries})`);
          initialRetries++;
          await new Promise((resolve) => setTimeout(resolve, 500));
          continue;
        }
        throw new Error(response.error || 'Failed to get job status');
      }

      const job = response.job;

      if (onProgress) {
        onProgress(job);
      }

      if (job.status === 'ready' || job.status === 'error') {
        return job;
      }

      await new Promise((resolve) => setTimeout(resolve, currentInterval));

      attempts++;

      if (attempts > 5 && currentInterval < 5000) {
        currentInterval = Math.min(currentInterval * 1.2, 5000);
      }
    }

    throw new Error('Flow diagram generation timed out');
  },

  /**
   * Delete a flow diagram job
   */
  async deleteJob(projectId: string, jobId: string): Promise<{ success: boolean; error?: string }> {
    try {
      const response = await axios.delete(
        `${API_BASE_URL}/projects/${projectId}/studio/flow-diagram-jobs/${jobId}`
      );
      return response.data;
    } catch (error) {
      if (axios.isAxiosError(error) && error.response) {
        return error.response.data;
      }
      log.error({ err: error }, 'failed to delete flow diagram job');
      throw error;
    }
  },
};

File: backend/app/api/studio/flow_diagrams.py
"""
Flow Diagram endpoints - AI-generated Mermaid diagrams.

Educational Note: Flow diagrams demonstrate visual representation of processes
and relationships using Mermaid.js syntax:
1. Claude analyzes source content
2. Identifies processes, workflows, or relationships
3. Creates appropriate Mermaid diagram syntax
4. Frontend renders using Mermaid library

Diagram Types:
- Flowcharts (graph TD/LR)
- Sequence diagrams
- State diagrams
- ER diagrams
- Class diagrams
- Pie charts, Gantt charts, etc.

Routes:
- POST   /projects/<id>/studio/flow-diagram              - Start generation
- GET    /projects/<id>/studio/flow-diagram-jobs/<id>    - Job status
- GET    /projects/<id>/studio/flow-diagram-jobs         - List jobs
- DELETE /projects/<id>/studio/flow-diagram-jobs/<id>    - Delete job
"""
import uuid
from flask import jsonify, request, current_app
from app.api.studio import studio_bp
from app.services.studio_services import studio_index_service
from app.services.studio_services.flow_diagram_service import flow_diagram_service
from app.services.source_services import source_index_service
from app.services.integrations.supabase import storage_service
from app.services.background_services.task_service import task_service
from app.services.auth import require_permission


@studio_bp.route('/projects/<project_id>/studio/flow-diagram', methods=['POST'])
@require_permission("studio", "flow_diagrams")
def generate_flow_diagram(project_id: str):
    """
    Start flow diagram generation or edit via background task.

    Educational Note: Flow diagrams are generated from source content using
    Claude to create Mermaid diagram syntax for visual process mapping.

    Request Body:
        - source_id: UUID of the source to generate diagram from (optional)
        - direction: Optional guidance for what to focus on
        - parent_job_id: UUID of the parent job to edit (optional, for edits)
        - edit_instructions: Instructions for editing the parent diagram (optional, for edits)

    Response:
        - success: Boolean
        - job_id: ID for polling status
        - message: Status message
    """
    try:
        data = request.get_json() or {}

        source_id = data.get('source_id')
        direction = data.get('direction', 'Create a diagram showing the key processes and relationships.')

        # Edit mode: load parent job's mermaid syntax as context for refinement
        parent_job_id = data.get('parent_job_id')
        edit_instructions = data.get('edit_instructions')
        previous_content = None

        if parent_job_id and not edit_instructions:
            return jsonify({
                'success': False,
                'error': 'edit_instructions is required when parent_job_id is provided'
            }), 400

        if parent_job_id:
            # Clean up any previously failed edit jobs for this parent
            try:
                all_jobs = studio_index_service.list_flow_diagram_jobs(project_id)
                for job in all_jobs:
                    if (job.get("status") == "error"
                            and job.get("parent_job_id") == parent_job_id):
                        studio_index_service.delete_flow_diagram_job(project_id, job["id"])
            except Exception:
                pass  # Non-critical cleanup -- don't block the new edit

            parent_job = studio_index_service.get_flow_diagram_job(project_id, parent_job_id)
            if not parent_job or not parent_job.get('mermaid_syntax'):
                return jsonify({
                    'success': False,
                    'error': 'Parent job not found or has no content to edit'
                }), 404

            previous_content = parent_job['mermaid_syntax']
            # Inherit source info from parent if not provided
            if not source_id:
                source_id = parent_job.get('source_id')

        # Source is optional — can generate from direction alone
        source_name = 'Direction Only'
        if source_id:
            source = source_index_service.get_source_from_index(project_id, source_id)
            if not source:
                return jsonify({
                    'success': False,
                    'error': f'Source not found: {source_id}'
                }), 404
            source_name = source.get('name', 'Unknown')

        # Create job record
        job_id = str(uuid.uuid4())
        studio_index_service.create_flow_diagram_job(
            project_id=project_id,
            job_id=job_id,
            source_id=source_id,
            source_name=source_name,
            direction=direction,
            parent_job_id=parent_job_id,
            edit_instructions=edit_instructions
        )

        # Submit background task
        task_service.submit_task(
            task_type="flow_diagram",
            target_id=job_id,
            callable_func=flow_diagram_service.generate_flow_diagram,
            project_id=project_id,
            source_id=source_id,
            job_id=job_id,
            direction=direction,
            previous_content=previous_content,
            edit_instructions=edit_instructions
        )

        return jsonify({
            'success': True,
            'job_id': job_id,
            'message': 'Flow diagram generation started',
            'source_name': source_name
        }), 202  # 202 Accepted - processing started

    except Exception as e:
        current_app.logger.error(f"Error starting flow diagram generation: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to start flow diagram generation: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/flow-diagram-jobs/<job_id>', methods=['GET'])
def get_flow_diagram_job_status(project_id: str, job_id: str):
    """
    Get the status of a flow diagram generation job.

    Response:
        - success: Boolean
        - job: Job record with status, progress, mermaid_syntax (when ready)
    """
    try:
        job = studio_index_service.get_flow_diagram_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Job not found: {job_id}'
            }), 404

        return jsonify({
            'success': True,
            'job': job
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error getting flow diagram job status: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to get job status: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/flow-diagram-jobs', methods=['GET'])
def list_flow_diagram_jobs(project_id: str):
    """
    List all flow diagram jobs for a project.

    Query Parameters:
        - source_id: Optional filter by source

    Response:
        - success: Boolean
        - jobs: List of job records
    """
    try:
        source_id = request.args.get('source_id')
        jobs = studio_index_service.list_flow_diagram_jobs(project_id, source_id)

        # Filter out orphaned failed-edit jobs (error + parent_job_id).
        # These are leftover from edit failures and should never be shown.
        clean_jobs = [
            job for job in jobs
            if not (job.get("status") == "error" and job.get("parent_job_id"))
        ]

        return jsonify({
            'success': True,
            'jobs': clean_jobs,
            'count': len(clean_jobs)
        }), 200

    except Exception as e:
        current_app.logger.error(f"Error listing flow diagram jobs: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to list jobs: {str(e)}'
        }), 500


@studio_bp.route('/projects/<project_id>/studio/flow-diagram-jobs/<job_id>', methods=['DELETE'])
def delete_flow_diagram_job(project_id: str, job_id: str):
    """
    Delete a flow diagram job and its files.

    Response:
        - Success status
    """
    try:
        job = studio_index_service.get_flow_diagram_job(project_id, job_id)

        if not job:
            return jsonify({
                'success': False,
                'error': f'Flow diagram job {job_id} not found'
            }), 404

        # Delete files from Supabase Storage
        storage_service.delete_studio_job_files(project_id, "flow_diagrams", job_id)

        # Delete job from index
        studio_index_service.delete_flow_diagram_job(project_id, job_id)

        return jsonify({
            'success': True,
            'message': f'Flow diagram job {job_id} deleted'
        })

    except Exception as e:
        current_app.logger.error(f"Error deleting flow diagram job: {e}")
        return jsonify({
            'success': False,
            'error': f'Failed to delete flow diagram job: {str(e)}'
        }), 500
